aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,680 @@
+#include <thrust/for_each.h>
+#include <thrust/functional.h>
+#include <thrust/gather.h>
+#include <thrust/generate.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/scan.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <thrust/system/cuda/execution_policy.h>
+#include <thrust/transform.h>
+#include <thrust/unique.h>
+
+#include <c10/cuda/CUDAMathCompat.h>
+
+namespace at {
+namespace native {
+namespace {
+
+// Number of threads in a block given an input size up to MAX_BLOCK_SIZE
+static int getNumThreads(int nElem) {
+#if defined(__HIP_PLATFORM_HCC__)
+  int threadSizes[5] = {16, 32, 64, 128, 256};
+#else
+  int threadSizes[5] = {32, 64, 128, 256, 512};
+#endif
+  for (int i = 0; i != 5; ++i) {
+    if (nElem <= threadSizes[i]) {
+      return threadSizes[i];
+    }
+  }
+  return threadSizes[4];
+}
+
+int64_t get_nvalues(const IntArrayRef& sizes, int64_t sparse_dim) {
+  /* Return the number of entries in the dense part of a sparse tensor.
+
+      `sizes` is a vector of sparse tensor dimensions.
+      `sparse_dim` is the dimension of the sparse part of a sparse tensor.
+    */
+  auto dim = sizes.size();
+  int64_t nvalues = 1;
+  for (auto i = sparse_dim; i < dim; i++) {
+    nvalues *= sizes[i];
+  }
+  return nvalues;
+}
+
+template <typename scalar_t, bool LogSoftMax>
+__global__ void cuda_sparse_coo_softmax_kernel(
+    int64_t* sorted_pool_indices,
+    int64_t size,
+    int64_t* pool_sizes,
+    int64_t* pool_offsets,
+    int64_t nvalues,
+    scalar_t* mx_rows,
+    scalar_t* exp_sums_rows,
+    PackedTensorAccessor<scalar_t, 2> input_values_acc,
+    PackedTensorAccessor<scalar_t, 2> ouput_values_acc) {
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
+    implementation of the sparse softmax algorithm that this implementation is
+    based on.
+  */
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int index = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  while (index < size) {
+    int64_t offset = pool_offsets[index];
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+    scalar_t* mx_row = mx_rows + index * nvalues;
+
+    /* Apply exp to (v - mx) and sum the results */
+    scalar_t* exp_sums_row = exp_sums_rows + index * nvalues;
+    for (int64_t p = 0; p < pool_indices_size; p++) {
+      auto i = pool_indices[p];
+      auto values_row = input_values_acc[i];
+      auto out_values_row = ouput_values_acc[i];
+      for (int64_t j = 0; j < nvalues; j++) {
+        auto v = c10::cuda::compat::exp(values_row[j] - mx_row[j]);
+        if (!LogSoftMax) {
+          out_values_row[j] = v;
+        }
+        exp_sums_row[j] += v;
+      }
+    }
+    for (int64_t j = 0; j < nvalues; j++) {
+      if (LogSoftMax) {
+        exp_sums_row[j] = c10::cuda::compat::log(exp_sums_row[j]);
```

Comment:
following the above, this should minimize writes to global memory by writing to a register (local) value first and then writing only the final values to global memory

Reply:
All right! I got it!

Reply:
Sorry to be so repetitive, we were just trying to mark ideas as we went through for clarity.

Reply:
Don't worry and thanks Mike, It nice to know  this  optimizations can be done. I've just tried to understand well until I got it the idea. 
So, I updated the PR covering most of the feedback, tomorrow I will update the benchmark and add new unittests. Thanks again :)